### ISSUE TITLE
Feature Request: Amplify Gen 2 - Define Global SelectionSets in resource.ts

### ISSUE THREAD

**Issue by ChristopherGabba** [aws-amplify/amplify-data#458]
### Is this related to a new or existing framework?

React Native

### Is this related to a new or existing API?

GraphQL API, New API

### Is this related to another service?

DynamoDB - Gen 2

### Describe the feature you'd like to request

Right now you have to define Selection Sets on your Gen 2 Queries each and every time you call them:
```typescript
      const response = await  client.models.Friendship.listFriendshipsByReceiverId(
          { receiverId: currentUser.userId },
          { selectionSet: friendshipSelectionSet },
        )
```

This also requires you define global selectionSet types like so:

```typescript
export const friendshipSelectionSet = [
  "id",
  "status",
  "senderId",
  "receiverId",
  "receiver.*", // want deeper hasMany relationship data included
  "sender.*", // want deeper hasMany relationship data included
  "createdAt",
  "updatedAt",
] as const

export type AWSFriendship = SelectionSet<
  Schema["Friendship"]["type"],
  typeof friendshipSelectionSet
>
```

But you may call this same type of query many times and you want the selectionSet the same every time.

### Describe the solution you'd like

I would like it to be as simple as defining it within the schema in backend.ts:
One option:
```typescript
  Friendship: a
    .model({
      id: a.id().required(),
      receiverId: a.id().required(),
      receiver: a.belongsTo("User", "receiverId").includeWithSelectionSet(), // added extra tag
      senderId: a.id().required(),
      sender: a.belongsTo("User", "senderId").includeWithSelectionSet(), // added extra tag
      status: a.ref("FriendStatus").required(),
    })
    .authorization((allow) => [allow.publicApiKey()])
    .secondaryIndexes((index) => [
      index("senderId")
        .name("bySender")
        .sortKeys(["receiverId"])
        .queryField("listFriendshipsBySenderId"),
      index("receiverId")
        .name("byReceiver")
        .sortKeys(["senderId"])
        .queryField("listFriendshipsByReceiverId"),
    ]),
```

Option 2:
```typescript
  Friendship: a
    .model({
      id: a.id().required(),
      receiverId: a.id().required(),
      receiver: a.belongsTo("User", "receiverId")
      senderId: a.id().required(),
      sender: a.belongsTo("User", "senderId")
      status: a.ref("FriendStatus").required(),
    })
    .authorization((allow) => [allow.publicApiKey()])
    .secondaryIndexes((index) => [
      index("senderId")
        .name("bySender")
        .sortKeys(["receiverId"])
        .queryField("listFriendshipsBySenderId"),
      index("receiverId")
        .name("byReceiver")
        .sortKeys(["senderId"])
        .queryField("listFriendshipsByReceiverId"),
    ])
  .selectionSet(["id",
      "status",
      "senderId",
      "receiverId",
      "receiver.*",
      "sender.*",
      "createdAt",
      "updatedAt"
]),
```

### Describe alternatives you've considered

Right now you have to define selectionSet on every query you want deeper data in and I just want it everywhere. 

### Is this something that you'd be interested in working on?

- [ ] 👋 I may be able to implement this feature request
- [ ] ⚠️ This feature might incur a breaking change

**Comment by stocaaro** [comment]
Hello @ChristopherGabba ,

I came across this reading through the backlog this morning and wanted to get a reply out.

I'm glad you mentioned the `as const` pattern, reading over your feature request I was thinking of that as the optimal pattern as things exist today. Thinking through this, if we add selection set specifications to the models themselves, what would it look like to remove these default selection sets when you don't need them? 

My concern with a feature like this is that your changing the request pattern in the client in a way that could be confusingly disconnected from the local call observed behavior. What I mean is that if I forgot I've done this and am seeing a large amount of data included in the responses when I don't expect it, I'm going to look at the client request, which would use your global variable today. If that connection is buried down the in the backend definition, it could take a lot longer to figure out as it depends on you knowing that this is a feature of the library data definition.

That said, I'm interested to hear how your app has come together. A year later, is this the pain point you thought it was? What are your thoughts on this with a bit of time?

@JanDoeTian ,

You're right that pagination through selection sets can be a bit awkward. Could you open an independent [feature request](https://github.com/aws-amplify/amplify-data/issues/new?template=feature_request.md) to capture this DX improvement?

Thanks,
Aaron

**Comment by ChristopherGabba** [comment]
@stocaaro Thanks for reaching out! I guess you would have way more experience with clients and how they use the amplify product. I've found that in my app in particular, I may call a `model.list` or a `model.create` and I use the same selectionSet everywhere. For me, the global selectionSet pattern would probably remove maybe 30 or 40 lines of code globally where I fetch data. 

That being said, I hear your concern. My opinion I guess, is that the 2nd option I mentioned above would probably be the most effective. Here is an example. Let's say I have a database param like:

```tsx
 User: a
    .model({
      id: a.id().required(),
      name: a.string().required(),
      birthdate: a.date(),
      markedForDelete: a.boolean(), // note here
    })
    .authorization((allow) => [allow.publicApiKey()])
```

and I have an account delete policy where if a user deletes their account, I flag this parameter as true. Then, I have a daily lambda function that runs and looks for accounts that have been `markedForDelete` for 3 days. 

I will never want this data in the client. Its strictly used as a backend flag. That means I have to call a custom `selectionSet` every single time in my app that I mention Users to prevent this `markedForDelete` data from coming to the client when I'm never going to use it. 

Defining a custom global selectionSet here is the only real solution. If I could, I would never have to call `selectionSet` on the graphql call.

**Comment by stocaaro** [comment]
> To your comment about a large amount of unexpected data coming to the client, I mean this happens for me by default because a lot of my stuff I don't need with every call.

Fair enough and well played. I wasn't thinking about using selection set to reduce the returned surface where I most commonly think to use it to bring in related models.

> in my app that I mention Users (which in my case is like 50 times lol) to prevent this markedForDelete data from coming to the client when I'm never going to use it.

Could this repetition be improved by capturing these calls in functions that receive the variant arguments and either default or override the selection set? something like:

```ts
type ListUserOptions = Parameters<typeof client.models.Todo.list>[0];

export function listUsers(options: ListUserOptions) {
  options.selectionSet ??= friendshipSelectionSet;
  client.models.User.list(options);
}
``` 
I imagine this living somewhere like `./src/lib/userApiCalls.ts` which  I could import and use everywhere my app makes calls to optimize out the awkward repetition.

I acknowledge that this is solving the problem in your code instead of the library. I think a library based solution might be to configure this in the client, something like looking to add: `client.models.User.defaultSelectionSet(friendshipSelectionSet);`, which would work much the same as the above internally. This has the same issue with separating the callsite behavior from the setting of the selection set, but keeping it in the client seems easier to support.

Interested to get your thoughts.

Thanks,
Aaron
